src/main/java/pl/edu/mimuw/LoanBankAction.java

```diff
@@ -1,9 +1,68 @@
 package pl.edu.mimuw;
 
-public abstract class LoanBankAction extends BankAction {
+import java.sql.Timestamp;
+import java.util.concurrent.TimeUnit;
 
-  // TODO:
-  // - remove abstract modifier from this class
-  // - this class has to contain extra fields other than DepositBankAction
-  // - you have to design it's constructors/accessors/setters and implement some from BankAction
+public class LoanBankAction extends BankAction {
+    private final double interestRate;
+    private final double monthlyDue;
+    private Timestamp lastPayment;
+    private int installments;
+
+    public LoanBankAction(double initial, Timestamp start,
+                          Timestamp end, String id, double interestRate) {
+        super(initial, start, end, id);
+        this.lastPayment = start;
+        this.interestRate = interestRate;
+        this.installments = (int) Math.max(getDepositPeriod() / 30, 1);
+        final var q = 1 + interestRate / installments;
+        final var x = Math.pow(q, installments);
+        this.monthlyDue = Math.round(initial * x * (q - 1) / (x - 1));
+    }
+
+    public boolean isEligible(BankClient client) {
+        return client.getBalance() >= monthlyDue;
+    }
+
+    public void execute(Bank bank, BankClient client) {
+        client.addBalance(initialAmount);
+        client.addAction(this);
+    }
+
+    public void collect(Bank bank, BankClient client) {
+        final var currentDate = bank.getCurrentDate();
+
+        if (installments > 0 && isPaymentDue(currentDate)) {
+            payInstallment(client);
+            lastPayment = currentDate;
+        }
+    }
+
+    public void payInstallment(BankClient client) {
+        client.subtractBalance(monthlyDue);
+        installments--;
+    }
+
+    // Deposit period in days.
+    private long getDepositPeriod() {
```

Comment:
metody pomocnicze mogą trafić do nadklasy abstrakcyjnej jako protected, to podklasy beda mogly tego uzywać